File: src/move/move.cpp
#include "move.h"

#include "../pieces/pieces.h"
#include "../vars.h"

auto Move::is_valid() const -> bool {
    Piece *start_piece = k_pieces[this->start],
          *end_piece = k_pieces[this->end];

    k_pieces[this->start] = nullptr;
    k_pieces[this->end] = start_piece;

    Square const king_pos = k_king_pos[k_current_player];

    if (typeid(*start_piece) == typeid(King))
        k_king_pos[k_current_player] = this->end;

    bool const valid =
        !dynamic_cast<King *>(k_pieces[k_king_pos[k_current_player]])
             ->is_checked();

    k_pieces[this->start] = start_piece;
    k_pieces[this->end] = end_piece;

    k_king_pos[k_current_player] = king_pos;

    return valid;
}

File: move.py
from dataclasses import dataclass
from typing import Optional

import piece
import square
import vars


@dataclass(frozen=True)
class Move:
    start: square.Square
    end: square.Square

    def is_valid(self) -> bool:
        start_piece: Optional[piece.Piece] = vars.PIECES[self.start]
        end_piece: Optional[piece.Piece] = vars.PIECES[self.end]
        king_pos = vars.KING_POS[vars.CURRENT_PLAYER]

        vars.PIECES[self.start], vars.PIECES[self.end] = (
            None, vars.PIECES[self.start]
        )

        if isinstance(start_piece, piece.King):
            vars.KING_POS[vars.CURRENT_PLAYER] = self.end

        valid: bool = not (
            vars.PIECES[vars.KING_POS[vars.CURRENT_PLAYER]].is_checked()
        )

        vars.PIECES[self.start], vars.PIECES[self.end] = start_piece, end_piece
        vars.KING_POS[vars.CURRENT_PLAYER] = king_pos

        return valid
